controller/service.go

```diff
@@ -144,6 +145,30 @@ func (c *controller) convergeBalancer(l log.Logger, key string, svc *v1.Service)
 		}
 	}
 
+	familyPolicy := v1.IPFamilyPolicySingleStack
+	if svc.Spec.IPFamilyPolicy != nil {
+		familyPolicy = *(svc.Spec.IPFamilyPolicy)
+	}
+
+	// If svc currently has 1 ip and policy PreferDualStack, try assigning ip from the missing family and same pool
+	if len(lbIPs) == 1 && familyPolicy == v1.IPFamilyPolicyPreferDualStack {
+		level.Info(l).Log("event", "tryAssignAdditionalIP", "msg", "familyPolicy is PreferDualStack, trying to assign additional ip")
+		existingFamily := ipfamily.ForAddress(lbIPs[0])
+		additionalFamily := ipfamily.IPv4
+		if existingFamily == ipfamily.IPv4 {
+			additionalFamily = ipfamily.IPv6
+		}
+		currentPool := c.ips.Pool(key)
+		newIP, err := c.ips.AllocateFromPoolForFamilyAndPolicy(key, svc, additionalFamily, currentPool, k8salloc.Ports(svc), SharingKey(svc), k8salloc.BackendKey(svc))
```

Comment:
Sure, I'll add the policy to make it clearer. Thanks for pointing it out!

I'm not sure if I understand your second point correctly, do you mean that when we use Assign, the ip that was assigned from the existingFamily will be lost? But we are only assigning the additional Family here?

Reply:
this in `Assign`:
```go
	alloc := &alloc{
		pool:  pool.Name,
		ips:   ips,
		ports: make([]Port, len(ports)),
		key:   *sk,
	}
	copy(alloc.ports, ports)
	a.assign(svcKey, alloc)
```
makes me think this would override the existing cache, without preserving the existing ip

Reply:
You mean in `preferDualStack` case? We do not use this function there, though?

Reply:
I added the comment in line 166 of service.go

Reply:
(1) Correct. `poolIps` will be filled with 1 ipv4 and 1 ipv6. Both are free, so the ipv4 is definitely not the one we assigned.

(2) We don't use the original family, but the family of the missing family, not the original ipfamily. In this case, the missing family would be ipv6, as we already had an ipv4. `selectIPsForFamilyAndPolicy` will return just the ipv6 part of the `poolIps`.

(3) The cache is only updated with ipv6, hence lbIPs will only have 2 ips.

Reply:
> We don't use the original family, but the family of the missing family. In this case, the missing family would be ipv6, as we already had an ipv4. selectIPsForFamilyAndPolicy will return just the ipv6 part of the poolIps.

but the policy passed is still "prefer", so `selectIPsForFamilyAndPolicy` will return both ips?

> The cache is only updated with ipv6, hence lbIPs will only have 2 ips.

iiuc `Assign` overwrites the cache with the `ips` it is given, so not sure how it is "only updates with ipv6", and lbIPs won't be with dupes

it seems there is some flaw with what's happening here, I'll write it up in a different comment